src/Cryptol/Utils/RecordMap.hs

```diff
@@ -0,0 +1,181 @@
+-- |
+-- Module      :  Cryptol.Utils.RecordMap
+-- Copyright   :  (c) 2020 Galois, Inc.
+-- License     :  BSD3
+-- Maintainer  :  cryptol@galois.com
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- This module implements an "order insensitive" datastructure for
+-- record types and values.  For most purposes, we want to deal with
+-- record fields in a canonical order; but for user interaction
+-- purposes, we generally want to display the fields in the order they
+-- were specified by the user (in source files, at the REPL, etc.).
+
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cryptol.Utils.RecordMap
+  ( RecordMap
+  , displayOrder
+  , recordFromFields
+  , recordFromFieldsErr
+  , recordFromFieldsWithDisplay
+  , canonicalFields
+  , displayFields
+  , lookupField
+  , adjustField
+  , traverseRecordMap
+  , mapRecordFields
+  , zipRecordsM
+  , zipRecords
+  , recordMapAccum
+  )
+where
+
+import           Control.DeepSeq
+import           Control.Monad.Except
+import           Data.Functor.Identity
+import           Data.Map (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Map.Merge.Strict as Map
+
+import Cryptol.Utils.Panic
+
+-- | An "order insensitive" datastructure.
+--   The fields can be accessed either according
+--   to a "canonical" order, or based on a
+--   "display" order, which matches the order
+--   in which the fields were originally specified.
+data RecordMap a b =
+  RecordMap
+  { recordMap :: !(Map a b)
+  , displayOrder :: [a]
```

Comment:
We should explicitly state the datatype invariant here, which I guess is that `keysSet recordMap == fromList displayOrder`. (But not in the haddock comment, as this is an abstract datatype.)